### ISSUE TITLE
mDNS not publishing when running in FreeBSD VNET jail

### ISSUE THREAD

**Issue by Wraul** [ikalchev/HAP-python#160]
So I had a lot of trouble with Home Assistant not showing up in the iOS home app when trying to pair.

I first tried the patched `accessory_driver.py` posted in this issue https://github.com/home-assistant/home-assistant/issues/17181
This did not improve things.
So I decided to do a bit of digging, and after a bit of experimental changes to the patched `accessory_driver.py` I managed to solve my issues.

I changed the intialization of `Zeroconf` from 

```python
self.advertiser = Zeroconf()
```
to
```python
self.advertiser = Zeroconf(InterfaceChoice.Default, False)
```

One must also import `InterfaceChoice` by changing
```python
from zeroconf import ServiceInfo, Zeroconf, NonUniqueNameException
```
to
```python
from zeroconf import ServiceInfo, Zeroconf, NonUniqueNameException, InterfaceChoice
```

So it appears that the default value of `InterfaceChoice.All` causes Zeroconf to fail to listen correctly.

Not 100% confident if the problem is with Zeroconf or with ifaddr which Zeroconf uses.
I'm reporting this here though if someone else encounters this, or if you want to work around the problem similar to how I did.

**Comment by cdce8p** [comment]
Something seems to be off with `zeroconf` and HomeKit. I've seen so many issue reports but couldn't reproduce them reliably. Just a moment ago when I was testing your suggestion @Wraul it decided not to work on the first pairing attempt. However the second one worked flawlessly, I didn't change anything in between (only a restart). I should mention that in my setup it works without any necessary adjustments.

**TLDR**: I don't know what to do with it anymore. Maybe you guys have an idea to get on top of it. Below my knowledge of these issues. Feel free to complement it where needed.

It seems there might be two separate issue related to `zeroconf` though, which start to crystallize:
 1. The accessory doesn't even show up in the `Add Accessory` screen when trying to pair.
 2. The accessory hangs after the pincode has been entered and woun't pair.

Some example issues:
https://github.com/home-assistant/home-assistant/issues/14567
https://github.com/home-assistant/home-assistant/issues/16929
https://github.com/home-assistant/home-assistant/issues/17181
https://github.com/home-assistant/home-assistant/issues/17372
https://community.home-assistant.io/t/another-post-about-pairing-with-homekit/68255

The problem with all of them is the various nature of this problem. Those include for

### Acc doesn't show up
 - System settings are wrong, network mode with docker, similar
 - Problem during the installation of or missing dependencies (obviously this is a Home Assistant problem that occurs sometimes but it complicates things never the less)
 - iOS devices not on the same local network
 - Already saved pairing in the `.accessory.state` file, so the services will be undiscoverable (This is especially an issue if combined with problem no.2
 - `Zeroconf` doesn't work properly???
 - `Zeroconf` update to `ifaddr` (with 0.21.0)???
 - (Something unknown??)

### Acc hangs during paring
 - `Zeroconf` throws a `NonUniqueNameException` during the first execution of `update_advertisement` when trying to register the updated service. Since the original one has been removed already `HomeKit` things the pairing didn't work and tries to restart. That in turn doesn't like `HAP-server` and sends a `403`.
Why `zeroconf` throws this exception I don't know. In those cases I had advised to use a modified version without the call to `update_advertisement` in `AccessoryDriver/pair`. That seems to work fine.

### Undiscovered other issues
Maybe

**Comment by cdce8p** [comment]
@Wraul What exactly has your problem been? At which point did you encounter issues?

**Comment by Wraul** [comment]
This was during my attempts to setup a new instance of Home Assistant in a FreeBSD 11.2 VNET jail.
The jail is connected to a VLAN but I don't think that has any impact.

No matter what I tried I couldn't get Home Assistant to show up for pairing.
I tried running `dns-sd -B _hap._tcp` on a nearby Mac and it showed that Home Assistant wasn't publishing the `_hap._tcp` mDNS entry.

I was confused though as `sockstat -l` showed one listener for `*:5353`.
Though after applying my fix it now shows two listeners for `*:5353` and Home Assistant shows up when running `dns-sd -B _hap._tcp`.

After applying the fix Home Assistant showed up in iOS and I could pair without any problems.

My theory is that ifaddr fails to iterate the interfaces correctly in the jail.

**Comment by cdce8p** [comment]
@Wraul can you try and downgrade `zeroconf` to version `0.20.0`? That would be the last version without `ifaddr`.

**Comment by Wraul** [comment]
@cdce8p I'm not 100% sure how I would go about doing that.
HAP-python is installed automatically by home assistant, with what I'm assuming are a fixed version.
I'm sure there is some way to work around that but I'm not that well versed in the Python ecosystem.

**Comment by cdce8p** [comment]
I don't know the details for your system, but usually `Home Assistant` is run in a python virtual environment. After HA is stopped, that python env needs to be activate in the console. These two commands should do the rest:
```bash
pip install --upgrade zeroconf==0.20.0

hass --skip-pip
```

**Comment by bdraco** [comment]
Also seems to cause duplicate connections from home hubs

**Comment by bdraco** [comment]
https://www.npmjs.com/package/multicast-dns seems to default to the same behavior so I'm not sure what the better default is.  While it solved the problem for me, it could cause breakage.

Ideally this is a configurable setting to All or Default

**Comment by bdraco** [comment]
With this change the "random" homed 100% cpu issue seems to have gone away as well.  Will test for a few days before coming up with a plan to move this forward
